File: lint-arwaky/src-python/capabilities/call_chain_analyzer.py
"""call_chain_analyzer — Call chain analysis capability."""

from __future__ import annotations
from ..taxonomy import (
    CallChainList,
    Count,
    DataFlowList,
    DirectoryPath,
    FileContentVO,
    FilePath,
    LineNumber,
    ResponseData,
    ResponseDataList,
    ScopeRef,
    SymbolName,
    SymbolNameList,
)

import re


from ..contract import (
    IDataFlowProtocol,
    IFileSystemPort,
    IScopeBoundaryResolverProtocol,
    ISemanticTracerProtocol,
    INamingVariantProtocol,
)


class CallChainAnalyzer(ISemanticTracerProtocol):
    """Call chain analyzer for JavaScript/TypeScript files."""

    def __init__(
        self,
        fs: IFileSystemPort,
        data_flow: IDataFlowProtocol,
        naming: INamingVariantProtocol,
        scope: IScopeBoundaryResolverProtocol | None = None,
    ):
        self._fs = fs
        self._data_flow = data_flow
        self._naming = naming
        self._scope = scope

    def get_variant_dict(self, name: SymbolName) -> ResponseData:
        return self._naming.get_variant_dict(name)

    def build_variants(self, name: SymbolName) -> SymbolNameList:
        return self._naming.build_variants(name)

    def get_enclosing_scope(
        self, file_path: FilePath, line: LineNumber
    ) -> ScopeRef | None:
        """Resolve enclosing scope using the scope resolver capability."""
        if not self._scope:
            return None
        return self._scope.resolve_enclosing_scope(file_path, line)

    def get_symbol_locations(
        self, file_path: FilePath, symbol: SymbolName
    ) -> ResponseDataList:
        """Stub for symbol location retrieval."""
        return ResponseDataList(values=[])

    def find_flow(
        self, file_path: FilePath, var_name: SymbolName, start_line: LineNumber
    ) -> DataFlowList:
        return self._data_flow.find_flow(file_path, var_name, start_line)

    def trace_call_chain(
        self, root_dir: DirectoryPath, target_name: SymbolName
    ) -> CallChainList:
        """Find all call sites for the target name within the project."""
        callers: list[SymbolName] = []
        name_str = str(target_name.value)
        call_pattern = re.compile(rf"\b{re.escape(name_str)}\s*\(")
        def_pattern = re.compile(rf"(?:function|class)\s+{re.escape(name_str)}\b")

        js_files: list[FilePath] = []
        for ext in ("*.js", "*.jsx", "*.ts", "*.tsx", "*.mjs"):
            js_files.extend(list(self._fs.glob(f"{root_dir}/**/{ext}")))

        for filepath in js_files:
            source_vo = self._fs.read_text(filepath)
            if not source_vo or not source_vo.value:
                continue

            file_lines = source_vo.value.splitlines()

            for i, line_str in enumerate(file_lines):
                if call_pattern.search(line_str) and not def_pattern.search(line_str):
                    rel_path = self._fs.get_relative_path(
                        filepath, DirectoryPath(value=str(root_dir))
                    )
                    call_site = f"{rel_path}:{i + 1} -> {line_str.strip()}"
                    callers.append(SymbolName(value=call_site))

        return CallChainList(values=callers)

    def project_wide_rename(
        self, root_dir: DirectoryPath, old_name: SymbolName, new_name: SymbolName
    ) -> Count:
        """Rename a symbol across all files in the project."""
        old_str = str(old_name.value)
        new_str = str(new_name.value)

        pattern = re.compile(
            rf"""
            (
                `(?:\\.|[^`\\])*`             |
                \"(?:\\.|[^\"\\])*\"          |
                '(?:\\.|[^'\\])*'             |
                //[^\n]*                      |
                /\*(?:.|\n)*?\*/
            )
            |
            \b({re.escape(old_str)})\b
            """,
            re.VERBOSE | re.DOTALL,
        )

        def _replacer(match: re.Match) -> str:
            if match.group(1) is not None:
                return match.group(1)
            return new_str

        js_files: list[FilePath] = []
        for ext in ("*.js", "*.jsx", "*.ts", "*.tsx", "*.mjs"):
            js_files.extend(list(self._fs.glob(f"{root_dir}/**/{ext}")))

        modified_count = 0
        for filepath in js_files:
            source_vo = self._fs.read_text(filepath)
            if not source_vo or not source_vo.value:
                continue

            source = source_vo.value

            if old_str in source:
                new_source = pattern.sub(_replacer, source)
                if new_source != source:
                    res = self._fs.write_text(filepath, FileContentVO(value=new_source))
                    if res.value:
                        modified_count += 1

        return Count(value=modified_count)

File: lint-arwaky/src-rust/capabilities/call_chain_analyzer.rs
// call_chain_analyzer — Call chain analysis capability for JS/TS files.
// Implements ISemanticTracerProtocol: trace_call_chain, project_wide_rename.

use std::fs;
use regex::Regex;
use crate::taxonomy::{DirectoryPath, FilePath, LineNumber, SymbolName, SymbolNameList};
use crate::capabilities::naming_variant_analyzer::NamingVariantAnalyzer;
use crate::capabilities::scope_boundary_resolver::ScopeBoundaryResolver;
use crate::capabilities::data_flow_analyzer::DataFlowAnalyzer;

const JS_EXTENSIONS: &[&str] = &[".js", ".jsx", ".ts", ".tsx", ".mjs"];

/// Call chain analyzer for JavaScript/TypeScript files.
pub struct CallChainAnalyzer {
    naming: NamingVariantAnalyzer,
    scope: ScopeBoundaryResolver,
    data_flow: DataFlowAnalyzer,
}

impl CallChainAnalyzer {
    pub fn new() -> Self {
        Self {
            naming: NamingVariantAnalyzer::new(),
            scope: ScopeBoundaryResolver::new(),
            data_flow: DataFlowAnalyzer::new(),
        }
    }

    /// Build all naming variants for a symbol.
    pub fn build_variants(&self, name: &SymbolName) -> SymbolNameList {
        self.naming.build_variants(name)
    }

    /// Resolve enclosing scope using the scope resolver capability.
    pub fn get_enclosing_scope(&self, file_path: &str, line: usize) -> Option<String> {
        self.scope.resolve_enclosing_scope(file_path, line)
    }

    /// Get data flow for a variable in a file.
    pub fn find_flow(&self, file_path: &str, var_name: &str, start_line: Option<usize>) -> Vec<String> {
        self.data_flow.find_flow(file_path, var_name, start_line)
            .into_iter()
            .map(|e| format!("Line {} [{}]: {}", e.line, e.kind, e.content))
            .collect()
    }

    fn collect_js_files(root_dir: &str) -> Vec<String> {
        let mut files = Vec::new();
        if let Ok(entries) = std::fs::read_dir(root_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.is_dir() {
                    let subfiles = Self::collect_js_files(&path.to_string_lossy());
                    files.extend(subfiles);
                } else if let Some(ext) = path.extension().and_then(|e| e.to_str()) {
                    let dot_ext = format!(".{}", ext);
                    if JS_EXTENSIONS.contains(&dot_ext.as_str()) {
                        files.push(path.to_string_lossy().to_string());
                    }
                }
            }
        }
        files
    }

    /// Find all call sites for the target name within the project.
    pub fn trace_call_chain(&self, root_dir: &str, target_name: &str) -> Vec<String> {
        let call_pattern = Regex::new(&format!(r"\b{}\s*\(", regex::escape(target_name))).unwrap();
        let def_pattern = Regex::new(&format!(r"(?:function|class)\s+{}\b", regex::escape(target_name))).unwrap();

        let js_files = Self::collect_js_files(root_dir);
        let mut callers: Vec<String> = Vec::new();

        for filepath in &js_files {
            let Ok(content) = fs::read_to_string(filepath) else { continue; };

            for (i, line) in content.lines().enumerate() {
                if call_pattern.is_match(line) && !def_pattern.is_match(line) {
                    // Compute relative path
                    let rel = filepath.strip_prefix(root_dir).unwrap_or(filepath).trim_start_matches('/');
                    let call_site = format!("{}:{} -> {}", rel, i + 1, line.trim());
                    callers.push(call_site);
                }
            }
        }

        callers
    }

    /// Rename a symbol across all JS/TS files in the project.
    pub fn project_wide_rename(&self, root_dir: &str, old_name: &str, new_name: &str) -> usize {
        let pattern = Regex::new(&format!(
            r#"(`(?:\\.|[^`\\])*`|"(?:\\.|[^"\\])*"|'(?:\\.|[^'\\])*'|//[^\n]*|/\*(?:.|\n)*?\*/)|(\b{}\b)"#,
            regex::escape(old_name)
        )).unwrap();

        let js_files = Self::collect_js_files(root_dir);
        let mut modified_count = 0;

        for filepath in &js_files {
            let Ok(source) = fs::read_to_string(filepath) else { continue; };
            if !source.contains(old_name) {
                continue;
            }

            let new_source = pattern.replace_all(&source, |caps: &regex::Captures| {
                if caps.get(1).is_some() {
                    caps[0].to_string()
                } else {
                    new_name.to_string()
                }
            }).to_string();

            if new_source != source {
                if fs::write(filepath, &new_source).is_ok() {
                    modified_count += 1;
                }
            }
        }

        modified_count
    }
}
